Implement the Happy Eye Balls RFC's

WyriHaximus opened the pull request:
By using the happy eye balls algorythm as descripted in [`RFC6555`](https://tools.ietf.org/html/rfc6555) and
[`RFC8305`](https://tools.ietf.org/html/rfc8305) it will connect to the quickest responding server with a
preference for IPv6.

WyriHaximus commented:
This PR is now ready for review. A follow up will be filed to replace the `DnsConnector` with the `HappyEyeBallsConnector` in the `Connector` an upcoming PR.

CharlotteDunois commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,287 @@
+    public function connect()
+    {
+        // skip DNS lookup / URI manipulation if this URI already contains an IP
+        if (false !== \filter_var($this->host, \FILTER_VALIDATE_IP)) {
```
```suggestion
        if (false !== \filter_var($this->uri, \FILTER_VALIDATE_IP)) {
```
:eyes:

WyriHaximus commented:
Ping @clue && @jsor did some major refactoring on this PR and it is now ready few review

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,288 @@
+        });
+    }
+
+    public function check()
```
All public methods should probably be marked `@internal`?

WyriHaximus replied:
That would make sense signalling that those methods are internal for the class (which is internal on it's own).

WyriHaximus replied:
done

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,288 @@
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            $that->resolve = $resolve;
+            $that->reject = $reject;
```
It's my understanding this is not safe when concurrent connection attempts are running?

WyriHaximus replied:
Hmm good catch, looking into it now :+1:

WyriHaximus replied:
Okey resolve this

WyriHaximus commented:
> @WyriHaximus Thank you very much for working on this, this is one of the features that is very high on my personal wishlist! 

:tada: !!!

> However, I realize that the implementation is anything but trivial. I've added some remarks to some places that look odd to me, perhaps you can take a look?

Thanks for the review, I've addressed those points.

> That being said, once all implementation issues are resolved, I wonder what's the best path forward to ensure a safe rollout of this feature in the future? Perhaps it makes sense to first enable this without IPv6 support to see the initial implementation does not break any existing implementations? I can see some (legacy) use cases where it may be desirable to explicitly disable IPv6 because the downstream application might not be IPv6-ready? (think database fields storing an address in a fixed length field)

Not entirely sure yet how to do that. What I want to do is write a follow up PR that initially will go full in on happy-eyeballs. And then we'll comes up during the discussion on that PR how we're shaping that. Because I'm pretty sure we can solve this with a flag on the `Connector` class :smile: .

clue commented on src/HappyEyeBallsConnector.php:
```diff
@@ -0,0 +1,37 @@
+    public function connect($uri)
+    {
+        try {
+            return HappyEyeBallsConnectionBuilder::create($this->loop, $this->connector, $this->resolver, $uri);
```
~It's my understanding this piece of logic never throws any exceptions, no?~ I see it will throw an `InvalidArgumentException` for invalid addresses, perhaps move this logic somewhere else to avoid vexing exceptions?

WyriHaximus replied:
Yup updated that and returning a rejected promise instead

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,304 @@
+            if (\count($connectQueueStash) > 0) {
+                $this->connectQueue[] = \array_shift($connectQueueStash);
+            }
+        }
```
I'm not sure I understand what I'm looking at tbqh.

WyriHaximus replied:
This mixes the new list of IP addresses (`$ips`) into the current one `$this->connectQueue`. This mixes the IPv4 addresses into the list of IPv6 addresses (or vise versa) in such a way that they alternate which IPvX is used to connect with.

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,304 @@
+                $connectionPromise->cancel();
+            }
+            unset($this->connectionPromises[$index]);
+            $connectionPromise = null;
```
What's going on with these `null` assignments? (see also below)

WyriHaximus replied:
Hehehe was attempting to hun/prevent memory leaks :D

WyriHaximus replied:
Cleaned those out

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,304 @@
+        // skip DNS lookup / URI manipulation if this URI already contains an IP
+        if (false !== \filter_var($this->host, \FILTER_VALIDATE_IP)) {
+            return $this->connector->connect($this->uri);
+        }
```
Perhaps move this logic closer to the beginning of the execution flow to skip this entire class?

WyriHaximus replied:
Done :+1:

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,304 @@
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
```
Interesting design decision, this really threw me off guard :grin:

WyriHaximus replied:
Same as above, to prevent memory leaks, the class symbolises a new connection attempt and will disappear once that connection has been established.

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,304 @@
+        $connectionBuilder = new self($loop, $connector, $resolver, $uri);
+
+        return $connectionBuilder->connect();
+    }
```
It looks like this class is only used internally anyway, so why not use the constructor directly?

WyriHaximus replied:
Done :+1:

WyriHaximus commented:
> @WyriHaximus Thanks for the update! This PR/implementation is not exactly trivial, so I've added some remarks to see if we can avoid some of its complexity. Let me know what you think about it.

It isn't trivial at all. I'll be adding links to certain implementation details linking to the RFC('s) where applicable to understand the why for those things.

WyriHaximus commented:
@clue oki updated to after your comments

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,295 @@
+            if ($typeHasBeenResolved === false) {
+                return false;
+
+                break;
```
leftover?

WyriHaximus replied:
Yup!

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,295 @@
+            $that->resolve(Message::TYPE_AAAA, $resolve, $reject);
+            Promise\Timer\resolve(0.05, $that->loop)->then(function () use ($that, $resolve, $reject) {
+                $that->resolve(Message::TYPE_A, $resolve, $reject);
+            });
```
It looks like this piece of logic is missing some cancellation handling, it should probably stop the timer or the lookup? What happens if I cancel the resulting promise? What happens if the IPv6 connection has succeeded already, it should probably stop the timer and not send the second lookup?

WyriHaximus replied:
It will still send the second look up. Just added cancelation logic

clue commented on src/HappyEyeBallsConnector.php:
```diff
@@ -0,0 +1,64 @@
+            return Promise\reject($exception);
+        } catch (\Throwable $throwable) {
+            return Promise\reject($throwable);
+        }
```
The try-catch looks like a leftover?

WyriHaximus replied:
Unless an `Error` is throw somewhere down the stack and we won't catch it then on PHP 7.

clue replied:
Not sure I follow, why would this piece of code throw an `Error` in the first place?

WyriHaximus replied:
It shouldn't, removed the try catch

WyriHaximus commented:
@clue working on the tests now, noticed the same thing this morning. 15% coverage to go :tada:

WyriHaximus commented:
@clue and we're at 100% :tada: !

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,278 @@
+            $that->resolverPromises[Message::TYPE_AAAA] = $that->resolve(Message::TYPE_AAAA, $resolve, $reject);
+            $that->resolverPromises[Message::TYPE_A] = Promise\Timer\resolve(0.05, $that->loop)->then(function () use ($that, $resolve, $reject) {
+                $that->resolverPromises[Message::TYPE_A] = $that->resolve(Message::TYPE_A, $resolve, $reject);
+            });
```
https://tools.ietf.org/html/rfc8305#section-3 appears to be a bit unclear, but to me it suggests both queries should be started simultaneously. If `AAAA` resolves first, connection attempts can start immediately. If `A` resolves first, it should wait 50ms for the `AAAA` response before starting connection attempts.

This appears to differ from the current implementation which always waits for 50ms to start the `A` query no matter what.

WyriHaximus replied:
Correct, my intention was to add that later on. But I suppose we can just as easily add it now

WyriHaximus replied:
IMHO I read that as send them both out as fast as possible after each other.

WyriHaximus replied:
Implemented

WyriHaximus commented:
> Implementing the Happy Eyeballs RFC is hard.
> 
> I've noticed some differences between this implementation and the algorithm as described in RFC 8305, perhaps you can take a look at this?

On it, some things have been left out because they are impossible tho.

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,289 @@
+                 */
+                return Promise\Timer\resolve(0.05, $that->loop)->then(function () use ($ips) {
+                    return Promise\resolve($ips);
+                });
```
Updated implementation looks much better to me!

The way I understand this piece of code is to now *always* wait for 50ms after the `A` DNS resolution before starting the first IPv4 connection. It's my understanding this should a maximum wait to at which point an IPv4 connection will be tried even when the `AAAA` query is still pending. Accordingly, if the `AAAA` query returns no results after 1ms, it should start trying IPv4 connections right after this response.

WyriHaximus replied:
Updated it to immediately after `AAAA` has been resolved release `A` for it's 50ms wait.

WyriHaximus commented:
@clue resolved the build issues.

WyriHaximus commented:
That's interesting because for me it's pretty consistent going for the IPv6 address:
![Screenshot from 2019-08-24 10-21-30](https://user-images.githubusercontent.com/147145/63634625-557aa200-c659-11e9-92ec-b0353aa46d3a.png)

Could you share how you tested this so I can run them from my end as well and compare results? Also the site in the screenshot is http://dual.tlund.se/ which I intend to use in the follow up PR as example/integration test.

These situations should be covered in the tests unless I missed something?

clue commented:
Here's a gist to add some dummy debugging logic (https://github.com/reactphp/socket/commit/cc1816000c6e1fa315b192c457e78756519a9a3a) and the unexpected output (happens at ~50% chance):

```
2019-08-24T11:12:17.830+02:00 connecting to clue.engineering
2019-08-24T11:12:17.832+02:00 found 2a03:4000:30:4bdc::11:9502
2019-08-24T11:12:17.832+02:00 trying 2a03:4000:30:4bdc::11:9502
2019-08-24T11:12:17.833+02:00 found 188.68.47.129
2019-08-24T11:12:17.833+02:00 trying 188.68.47.129
2019-08-24T11:12:17.850+02:00 yey 188.68.47.129
2019-08-24T11:12:17.850+02:00 nay 2a03:4000:30:4bdc::11:9502
2019-08-24T11:12:17.850+02:00 tcp://192.168.178.48:44060=>tcp://188.68.47.129:80
```

The timestamps suggests that IPv6 resolves first (OK), IPv6 connection attempt is started (OK), IPv4 resolves next (OK), IPv4 connection attempt is immediately started (nay?), IPv4 connection succeeds first (OK), IPv6 connection is rejected (OK), IPv4 connection wins (OK).

Unless I'm missing something, the fact that it tries both stacks at the same time seems to be in conflict with https://tools.ietf.org/html/rfc8305#section-5 which suggests a "connection attempt delay" of 100-250ms. This implies that in the above output, the IPv4 connection should be started slightly delayed (unless the IPv6 connection fails earlier, in which case it can start the next attempt right away). What do you think about this?

WyriHaximus commented:
Here are my test results, I ran it against `clue.engineering`, `dual.tlund.se`, `wyrihaximus.net`, and `google.com` 100 times with a 1 second grace period between it. Using:

```
(for i in `seq 1 100`; do php examples/11-http-client.php google.com | grep 2019-08-24; done) >> google.com.txt
```

In summary `clue.engineering` connected over IPv4 99% of the connections, and only 1% over IPv6. Interestingly enough `dual.tlund.se` was the complete opposite. Both `wyrihaximus.net`, and `google.com` connected over IPv6 100% of the time. I've added miliseconds to the timestamp and can confirm the 50ms wait after IPv6 resolved and tried to connect before trying to connect over IPv4. Another thing I've noticed is that it only got a single one, tried `blog.wyrihaximus.net` and it jumped to IPv6 insanely fast:

```
2019-08-24T13:35:48.261+02:00.261674 connecting to blog.wyrihaximus.net
2019-08-24T13:35:48.266+02:00.266716 found 2600:9000:21b3:9000:5:3bd8:cc80:93a1, 2600:9000:21b3:2e00:5:3bd8:cc80:93a1, 2600:9000:21b3:aa00:5:3bd8:cc80:93a1, 2600:9000:21b3:9600:5:3bd8:cc80:93a1, 2600:9000:21b3:8c00:5:3bd8:cc80:93a1, 2600:9000:21b3:4400:5:3bd8:cc80:93a1, 2600:9000:21b3:600:5:3bd8:cc80:93a1, 2600:9000:21b3:ce00:5:3bd8:cc80:93a1
2019-08-24T13:35:48.266+02:00.266764 trying 2600:9000:21b3:9000:5:3bd8:cc80:93a1
2019-08-24T13:35:48.267+02:00.267449 found 143.204.101.127, 143.204.101.62, 143.204.101.72, 143.204.101.5
2019-08-24T13:35:48.280+02:00.280764 yey 2600:9000:21b3:9000:5:3bd8:cc80:93a1
2019-08-24T13:35:48.280+02:00.280815 tcp://[::1]:46664=>tcp://[2600:9000:21b3:9000:5:3bd8:cc80:93a1]:80
```


Test results:

[clue.enginering.txt](https://github.com/reactphp/socket/files/3537137/clue.enginering.txt)
[dual.tlund.se.txt](https://github.com/reactphp/socket/files/3537138/dual.tlund.se.txt)
[google.com.txt](https://github.com/reactphp/socket/files/3537139/google.com.txt)
[wyrihaximus.net.txt](https://github.com/reactphp/socket/files/3537140/wyrihaximus.net.txt)

clue commented:
What we're seeing seems to be the same, but with slightly different timings (most likely due to different routing / geographical locations). Your logging output confirms that the second connection will be started while the first connection attempt is still pending. As per the above RFC section, this should only happen after waiting at least 100ms.

Snippet from your last log files:

```
2019-08-24T12:53:44.778+02:00.778011 connecting to wyrihaximus.net
2019-08-24T12:53:44.790+02:00.790520 found 2a03:b0c0:0:1010::ae:8001
2019-08-24T12:53:44.790+02:00.790724 trying 2a03:b0c0:0:1010::ae:8001
2019-08-24T12:53:44.795+02:00.795194 found 95.85.34.55
2019-08-24T12:53:44.795+02:00.795398 trying 95.85.34.55 <-- this one starts 5ms after the first, should wait 100ms
2019-08-24T12:53:44.800+02:00.800781 yey 2a03:b0c0:0:1010::ae:8001
2019-08-24T12:53:44.801+02:00.801329 nay 95.85.34.55
2019-08-24T12:53:44.803+02:00.803024 tcp://[::1]:51268=>tcp://[2a03:b0c0:0:1010::ae:8001]:80
```

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,311 @@
+
+                    if ($that->timer instanceof TimerInterface) {
+                        return;
+                    }
```
This check causes the bug discussed above. The timer will only be started when more IPs are available when the first connection attempt (IPv6) is started. If the IPv4 address resolves afterwards, then this timer is not running and it will immediately try the IPv4 connection instead of waiting first.

clue replied:
Can you take a look at this again? I can't reproduce this locally anymore, but I don't see any change here?

WyriHaximus replied:
This was solved in the check method.

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,311 @@
+                 * @link https://tools.ietf.org/html/rfc8305#section-3
+                 */
+                $ipv4Deferred = new Promise\Deferred();
+                $timer = Promise\Timer\resolve(0.05, $that->loop);
```
There's no clean up logic in case the resulting promise is cancelled while this timer is running. I would suggest directly using the loop instead.

WyriHaximus commented:
Well partially, the 50ms (will dive into why it only waits 5ms) is the wait for IPv6 after IPv4 address already resolved before using IPv4. But that 100ms should also kick in ensuring it follows the RFC. Will check your pointers out and fix this issue. Yeah IPv6 and IPv4 routes aren't guaranteed to be the same so an IPv6 route my actually be a lot slower then an IPv4 route or viseversa.

Here is a nice overview: https://tools.ietf.org/html/rfc8305#section-8

WyriHaximus commented:
Ok just fixed the bug, been using `www.govt.nz` to test against because of the latency and observed the proper delay between connections:

```
2019-08-24T19:13:41.662+02:00.662217 connecting to www.govt.nz
2019-08-24T19:13:41.675+02:00.675571 found 2a02:e980:1b::bb, 2a02:e980:1d::bb
2019-08-24T19:13:41.675+02:00.675882 trying 2a02:e980:1b::bb
2019-08-24T19:13:41.680+02:00.680684 found 103.28.251.187, 103.28.250.187
2019-08-24T19:13:41.772+02:00.772988 trying 103.28.251.187
2019-08-24T19:13:41.873+02:00.873665 trying 2a02:e980:1d::bb
2019-08-24T19:13:41.964+02:00.964694 yey 2a02:e980:1b::bb
2019-08-24T19:13:41.965+02:00.965516 nay 103.28.251.187
2019-08-24T19:13:41.967+02:00.967636 nay 2a02:e980:1d::bb
2019-08-24T19:13:41.969+02:00.969792 tcp://[::1]:51742=>tcp://[2a02:e980:1b::bb]:80
```

clue commented:
My debug output doesn't match with what you're seeing. I see the test suite wasn't updated with your last changes, perhaps this is something we can start with?

```
2019-08-25T09:27:55.252+02:00 connecting to clue.engineering
2019-08-25T09:27:55.254+02:00 found 2a03:4000:30:4bdc::11:9502
2019-08-25T09:27:55.254+02:00 trying 2a03:4000:30:4bdc::11:9502
2019-08-25T09:27:55.255+02:00 found 188.68.47.129
2019-08-25T09:27:55.255+02:00 trying 188.68.47.129 <-- this starts 1ms after the IPv6 connection, it should wait at least 100ms!
2019-08-25T09:27:55.272+02:00 yey 2a03:4000:30:4bdc::11:9502
2019-08-25T09:27:55.272+02:00 nay 188.68.47.129
2019-08-25T09:27:55.272+02:00 tcp://[2003:c9:170d:2600:6097:37d6:3cbf:8a5a]:48632=>tcp://[2a03:4000:30:4bdc::11:9502]:80
```

We should assert that a second connection will not be attempted while the first one is pending for less than 0.1 seconds. We should assert that this "next connection" timer is started. We should assert that a second connection attempt is started when the first attempt is still pending, but the timer fired. Likewise, add similar tests to check variations of IPv4/IPv6 attempts depending on DNS resolution etc.

WyriHaximus commented:
> We should assert that a second connection will not be attempted while the first one is pending for less than 0.1 seconds. 

✅

> We should assert that this "next connection" timer is started.

✅

> We should assert that a second connection attempt is started when the first attempt is still pending, but the timer fired.


✅
> Likewise, add similar tests to check variations of IPv4/IPv6 attempts depending on DNS resolution etc.

❌ (working on it)

WyriHaximus commented:
> Likewise, add similar tests to check variations of IPv4/IPv6 attempts depending on DNS resolution etc.

✅

clue commented on src/HappyEyeBallsConnectionBuilder.php:
```diff
@@ -0,0 +1,317 @@
+                $ipv4Deferred = new Promise\Deferred();
+                $deferred = new Promise\Deferred();
+
+                $timer = $that->loop->addTimer($that::RESOLVE_WAIT, function () use ($deferred, $ips) {
```
Promise cancellation doesn't seem to cancel this timer?

WyriHaximus replied:
Addressed ghag

clue commented on tests/HappyEyeBallsConnectorTest.php:
```diff
@@ -0,0 +1,539 @@
+            if (count($ipv4) > 0) {
+                $this->tcp->expects($this->at($i++))->method('connect')->with($this->equalTo('scheme://' . array_shift($ipv4) . ':80/?hostname=google.com'))->will($this->returnValue(Promise\resolve()));
+            }
+        }
```
Not a big fan of having this logic in the test suite. It's my understanding this should be rather static and/or multiple independent tests. After all, where are the tests for this test suite? :eyes: I'm okay with this for now.

WyriHaximus replied:
Not a big fan of it either. But this lets us test different lenghth IPvX responses against several tests. This uncovered certain edge cases in the past and help me first the issue you commented about on line 60.

WyriHaximus commented:
@clue decreased tests time, they're still slower then before but that's due to the timings involved.

clue commented:
@WyriHaximus Can confirm the test time is now down to ~50s on my machine. There still seem to be plenty of timers, so I would still suggest mocking the loop and explicitly invoking the timer callback like this:

```php
$loop = $this->createMock(LoopInterface::class);
$timer = null;
$loop->expects($this->once())->method('addTimer')->with(2.0, $this->callback(function ($cb) use (&$timer) {
    $timer = $cb;
    return true;
}));

// continue with test

$this->assertNotNull($timer);
$timer();
```

WyriHaximus commented:
@clue yeah this is going to be `fun` :zipper_mouth_face:

WyriHaximus commented:
@clue went for a slightly different approach that still does a full run with an actual event loop and timers, but I've speed it up instead by a factor of 10.

WyriHaximus commented on README.md:
```diff
@@ -1154,6 +1155,60 @@ be used to set up the TLS peer name.
+It will then replace the hostname in the destination URI with this IP's and
+append a `hostname` query parameter and pass this updated URI to the underlying
+connector. 
+The Happy Eye Balls algorythm describes looking the IPv6 and IPv4 address for 
```
`algorithm`

WyriHaximus commented on README.md:
```diff
@@ -1154,6 +1155,60 @@ be used to set up the TLS peer name.
+The `HappyEyeBallsConnector` class implements the
+[`ConnectorInterface`](#connectorinterface) and allows you to create plaintext
+TCP/IP connections to any hostname-port-combination. Internally it implements the 
+happy eyeballs algorythm from [`RFC6555`](https://tools.ietf.org/html/rfc6555) and 
```
`algorithm`

WyriHaximus commented:
So @jsor spotted two typo's after approving, will fix those before merging 👍
